File: spellbound-ngram/spellbound-ngram-scala/src/main/scala/net/xrrocha/scala/spellbound/ngram/StringMetric.scala
package net.xrrocha.scala.spellbound.ngram

import info.debatty.java.stringsimilarity.Damerau
import org.apache.lucene.search.spell.{JaroWinklerDistance, LevensteinDistance}

trait StringMetric {
  def maxDistance: Double

  def stringDistance(s1: String, s2: String): Double
}

object DamerauStringMetric extends StringMetric {
  lazy val maxDistance = .275
  lazy val damerau = new Damerau

  def stringDistance(s1: String, s2: String): Distance = {
    val maxLength = math.max(s1.length, s2.length)
    1.0 - (maxLength - damerau.distance(s1, s2)) / maxLength
  }
}

trait LuceneStringMetric extends StringMetric {
  def stringDistanceMetric: org.apache.lucene.search.spell.StringDistance

  def stringDistance(s1: String, s2: String): Double =
    1.0 - stringDistanceMetric.getDistance(s1, s2)
}

object LevenshteinStringMetric extends LuceneStringMetric {
  lazy val maxDistance = .75
  lazy val stringDistanceMetric = new LevensteinDistance
}

object JaroWinklerStringMetric extends LuceneStringMetric {
  lazy val maxDistance = .85
  lazy val stringDistanceMetric = new JaroWinklerDistance
}

trait StringMetricConfig extends ConfigSettings {
  self: StringMetric =>

  lazy val configName = "stringMetric"

  lazy val maxDistance: Double = config.getDouble("maxDistance")

  lazy val stringMetric: StringMetric =
    if (!config.hasPath("stringMetricType")) {
      DamerauStringMetric
    } else {
      lazy val stringMetricName: String = config.getString("stringMetricType")
      stringMetricName.toLowerCase match {
        case "damerau" => DamerauStringMetric
        case "levenstein" => LevenshteinStringMetric
        case "jarowinkler" => JaroWinklerStringMetric
        case _ =>
          Class.forName(stringMetricName).asInstanceOf[StringMetric]
      }
    }
}

File: spellbound-ngram/spellbound-ngram-kotlin/src/main/kotlin/net/xrrocha/kotlin/spellbound/ngram/StringMetric.kt
package net.xrrocha.kotlin.spellbound.ngram

import org.apache.lucene.search.spell.JaroWinklerDistance
import org.apache.lucene.search.spell.LevensteinDistance
import org.apache.lucene.search.spell.StringDistance

interface StringMetric {
  fun stringSimilarity(s1: String, s2: String): Double
}

interface LuceneStringMetric : StringMetric {
  val stringDistance: StringDistance

  override fun stringSimilarity(s1: String, s2: String) = stringDistance.getDistance(s1, s2).toDouble()
}

object LevensteinStringMetric : LuceneStringMetric {
  override val stringDistance = LevensteinDistance()
}

object JaroWinklerStringMetric : LuceneStringMetric {
  override val stringDistance = JaroWinklerDistance()
}
